Cabal Init Omnibus

emilypi opened the pull request:
---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

## New additions

This PR does the following: 

- Completely rewrites the `cabal init` command to have a saner structure that can be tested easily. 
- Optimizing for individual component target generation allows us to golden test individual stanzas under various scenarios instead of just whole cabal files
- Coverage reports may now be generated for every component of the command. Coverage works for the whole project now, full stop. 
- Control flow for interactive vs. non-interactive vs. simple target generation is completely decoupled.
- Cleans up the file creation spaghetti, and prunes unused or redundant behavior wherever it arises. 
- Rewrites the prompt code to be less unnecessarily typed (strings are fine!)

This comes with bugfixes: 
 
- [x] Fixes #7273 
- [x] Fixes #7253 
- [x] Fixes #7256 
- [x] Fixes #7255  
- [x] Fixes #7251
- [x] Fixes #6758 
- [x] Fixes #6864 
- [x] Fixes #1074

phadej commented on Cabal/src/Distribution/Fields/Pretty.hs:
```diff
@@ -35,6 +35,7 @@ import qualified Text.PrettyPrint as PP
 data PrettyField ann
     = PrettyField ann FieldName PP.Doc
     | PrettySection ann FieldName [PP.Doc] [PrettyField ann]
+    | PrettyEmpty
```
Why this change?

phadej replied:
I'd suggesting trying whether using `Maybe (PrettyField ann)` and collecting `[Maybe (PrettyField ann)]` and `catMaybes` that into sections.

Another option is to have

```haskell
    | PrettyEmpty ann
```

so 

```haskell
prettyFieldAnn :: PrettyField ann -> ann
```

is definable. I.e. every constructor of `PrettyField` can be annotated.

(There is such function for parser `Field`: https://hackage.haskell.org/package/Cabal-3.4.0.0/docs/Distribution-Fields-Field.html#v:fieldAnn)

emilypi replied:
I like the `PrettyEmpty ann` suggestion. The use of `Maybe` here is what i'm trying to avoid, since we must accept `Maybe` returns and the use of `catMaybes` in every bit of code that interfaces with the formatter, when it's such a small and useful addition to just add directly in the printer.

emilypi replied:
it occurred to me that we can still write `prettyFieldAnn`, we just have to choose a sensible annotation default. That is, we would not have `PrettyField ann -> ann`, but rather 

```haskell
prettyFieldAnn :: Default ann => PrettyField ann -> ann
```

Or `Monoid ann` which could provide a mempty in the case of the printer. Just adding that case for `PrettyEmpty` traded in 20 lines of code in 3 modules for 3 lines of code in 1 module, and I have less finicky data prep to be concerned with in the stanza creation. I'm inclined to keep it as-is.

phadej commented:
Why bundle coverage changes and cabal init changes? Are they related? I don't see how.

phadej commented on cabal-install/cabal-install.cabal:
```diff
@@ -20,60 +17,38 @@ Copyright:          2003-2020, Cabal Development Team
-      if impl(ghc < 8.8)
-        ghc-options: -Wnoncanonical-monadfail-instances
 
+library
```
There are relevant issues, like
- https://github.com/haskell/cabal/issues/6090
- https://github.com/haskell/cabal/issues/7224
- https://github.com/haskell/cabal/issues/3781

Is this trying to address them?

emilypi replied:
This is actually in service of coverage reporting, but if it fixes them, great.

phadej commented:
I don't like how this contains at least three separate changes in one PR. I hope this is just a draft.

emilypi commented:
@phadej it's clearly marked a draft. I'll split it out into several PRs if that makes you less uneasy.

emilypi commented:
@fgaz just so @ptkato and I can parallelize a bit, how do you want to split this PR up into reviewable chunks?

fgaz commented:
At the commit level, ideally the smallest chunks that make sense (and build) on their own. At the pr level, strongly related changes can be on the same pr. Or not. So something like this, following OP:

* PR: infrastructure stuff (should probably be a single pr since most of the stuff depends on the previous)
  * commit: Expose modules as part of a cabal-install library
  * commit: Gets rid of the dogfooding framework...
  * commit: generate coverage reports
  * commit: Move cabal-install-solver out to its own directory.
  * commit: Move all long-running tests to their own test target
* PR: fix for HPC generation (or was this the pr that just got merged?)
* PR: init stuff
  * no idea if this can be split. probably not?

emilypi commented:
> fix for HPC generation (or was this the pr that just got merged?)

Yes, i depended on his PR to get us multiproject support and generate coverage for the tests we're writing in this PR. 

> PR: infrastructure stuff (should probably be a single pr since most of the stuff depends on the previous)

Okay so piggy-backing on those suggestions, in terms of a timeline, i think this sequence is the one we need:

1.  Expose `cabal-install` library + remove dogfooding (these are dependent upon each other for testing)
2.  Split out the `cabal-install-solver` package
3.  Split tests out into unit-tests and long-running test targets (prerequisite)
4.  Change to the `cabal.project` setup for coverage generation 
5. Init work

emilypi commented on cabal-install/tests/fixtures/init/golden/cabal/cabal-lib-and-exe-with-comments.golden:
```diff
@@ -80,31 +76,36 @@ executable foo
-        base ^>=4.13.0.0,
-        containers ^>=5.7.0.0,
-        unordered-containers ^>=2.7.0.0
+        base,
```
FYI, these `base` dependencies are artificially mangled so that we can check that the golden files are generated homogeneously across GHC versions. There are other unit tests that cover the correctness of the `base` dependency version inferrance, and the user will see `base ^>=4.14.0.0` etc. in practice. 

As they are, the original golden files are actually _incorrect_ for any GHC < 8.8 in more ways than one: note the version commentary generated above. It's off by 1!

Mikolaj replied:
Is this comment resolved? It's the only one I can see in https://github.com/haskell/cabal/pull/7344/files view.

emilypi replied:
Yes this is resolved.

emilypi commented on cabal-install/src/Distribution/Client/Init/Interactive/Command.hs:
```diff
@@ -69,11 +70,12 @@ import Language.Haskell.Extension (Language(..))
     -> InitFlags
     -> m ProjectSettings
-createProject v pkgIx srcDb initFlags = do
+createProject v _comp pkgIx srcDb initFlags = do
```
This is unused in 2 cases (interactive + simple). I'd just configure `NonInteractive.createProject` to take the compiler arg first and then return `NonInteractive.createProject comp` as the function returned in the where clause.

Mikolaj commented on tests/fixtures/init/golden/pkg-desc/pkg-with-comments.golden:
```diff
@@ -0,0 +1,46 @@
+build-type:         Simple
+
+-- Extra files to be distributed with the package, such as examples or a README.
+extra-source-files: CHANGELOG.md
```
The PR supposedly fixes #6864, but this doesn't list the changelog as `extra-doc-files`. How so?

emilypi replied:
We're finding a few bugs in a few places like that as we go along with our internal review. It seems we marked "ready to review" a little prematurely. I'll ping when we have this fixed (it's simple fix)

Mikolaj replied:
I'm still seeing this.

emilypi replied:
Where are you still seeing this? It doesn't match the golden files I'm seeing: https://github.com/haskell/cabal/blob/76468fbedbbc92e613fceaf43057de24caeebc1a/cabal-install/tests/fixtures/init/golden/pkg-desc/pkg-with-comments.golden

Mikolaj replied:
I think github was lying to me. I see the problem no longer. :)

emilypi replied:
I actually found the problem (and it was a problem!): check out the file name. It's at the root of the project! Golden files must have been run from the root using `cabal run cabal-install:unit-tests`, which creates the golden files at whatever the current directory is :angry:

I've cleared out the stale golden files.

Mikolaj reviewed (approved):
Awesome mega-PR. I can't do it justice, not taking part in the epic struggle from the beginning, but I'm duly impressed by its breadth and the law and order it imposes. I would feel uncomfortable with such a large PR, but I understand it provides an excellent test coverage for what it changes. Great job.

emilypi commented:
:tada:
